### PR TITLE
Fix ASan container-overflow in ap_probe_expansion

### PR DESCRIPTION
**Pull request by captain5050**
Defer erasure of probes in `SessionExpander` to avoid modifying `ast_.root->probes` while iterating over it.

Previously, `SessionExpander::visit(Probe &)` would call `std::erase` on `ast_.root->probes` immediately when a matching retprobe was found. Since this visitation happens during an iteration over `ast_.root->probes` in the base `Visitor`, modifying the vector invalidated iterators and led to a sanitizer detected `container-overflow` when the range-based for loop continued to its cached end.

Now, `SessionExpander` collects probes to be erased and performs the erasure in an overridden `visit(Program &)` method after the base visitation completes.

Co-developed-by: Gemini

##### Checklist

- [ x] Language changes are updated in `docs/language.md`, `docs/stdlib.md`, or `man/adoc/bpftrace.adoc`
- [ x] User-visible and non-trivial changes updated in `CHANGELOG.md`
- [ x] The new behaviour is covered by tests

### CODE REVIEW

**Comment by jordalgo** [comment]
@captain5050 Thanks for the first contribution. There is an issue here (which is untested) in that because we're not deleting the return probe right away multiple entry probes could potentially match against 1 return probe, e.g. `kprobe:func_* { @a = 1 } kprobe:func_* { @b = 1 } kretprobe:func_* { @r = 1 }` causing us to lose the events of the 2nd duplicate entry probe. This is an edge case in practice but we should guard against this occurrence and ensure a 1 to 1 match. 

Mind also adding a unit test for this when you make the update?

**Comment by captain5050** [comment]
> @captain5050 Thanks for the first contribution. There is an issue here (which is untested) in that because we're not deleting the return probe right away multiple entry probes could potentially match against 1 return probe, e.g. `kprobe:func_* { @a = 1 } kprobe:func_* { @b = 1 } kretprobe:func_* { @r = 1 }` causing us to lose the events of the 2nd duplicate entry probe. This is an edge case in practice but we should guard against this occurrence and ensure a 1 to 1 match.

Hmm.. the matching could skip when probes_to_erase_ contains the same probe, to keep the existing behavior.

> Mind also adding a unit test for this when you make the update?

So I was seeing the failure with tests/ap_probe_expansion.cpp do you have any suggestions on what/where to add additional testing?

**Comment by jordalgo** [comment]
> So I was seeing the failure with tests/ap_probe_expansion.cpp do you have any suggestions on what/where to add additional testing?

Just thinking of an additional unit test similar to `TEST(ap_probe_expansion_btf, kprobe_session)` where you validate that if there are two kprobes with the same attachpoint that only one of them gets turned into a session probe (when there is only one matching kretprobe)

File: src/ast/passes/ap_probe_expansion.cpp [hunk 1]
```diff
@@ -167,10 +169,14 @@ Probe *SessionExpander::find_matching_retprobe(Probe &probe)
...
-
-  return nullptr;
+  if (retprobes.size() != 1) {
+    return nullptr;
+  }
+  // Return the probe as long as it isn't scheduled for erasure.
+  auto begin = probes_to_erase_.begin();
+  auto end = probes_to_erase_.end();
+  auto found = std::find(begin, end, retprobes[0]);
+  return found == end ? retprobes[0] : nullptr;
```
Comment (jordalgo) [inline src/ast/passes/ap_probe_expansion.cpp:179]: nit: let's just move this logic into the big copy_if return above `std::find(probes_to_erase_.begin(), probes_to_erase_.end(), other) == probes_to_erase_.end()`

**Review by jordalgo — CHANGES_REQUESTED** [review]

**Comment by jordalgo** [comment]
@captain5050 Just curious, any interest in finishing this out?

**Comment by jordalgo** [comment]
Landing this instead: https://github.com/bpftrace/bpftrace/pull/5175
